Fix shardLeader cache concurrent access

congqixia opened the pull request:
Fix  write map without mutex control
Also GetShards returns a copy of leader list instead of original one

Related to #17072

/kind bug

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

XuanYang-cn reviewed (changes requested):
The return of `updataShardsWithRounRobin` can be dropped @congqixia

congqixia commented:
> The return of `updataShardsWithRounRobin` can be dropped @congqixia

@XuanYang-cn Done and updated

congqixia commented:
/assign @XuanYang-cn

XuanYang-cn commented:
/lgtm

congqixia committed:
Fix shardLeader cache concurrent access

Fix  write map without mutex control
Also GetShards returns a copy of leader list instead of original one

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

XuanYang-cn commented:
/lgtm
